cc_ssh_import_id: support space separated list of users

raharper opened the pull request:
Add support for parsing a space separated list of users. This adds
to a single string which becomes a list, and supports platform
consoles which cannot utilize [] keys for building yaml list.

igalic commented:
> supports platform consoles which cannot utilize [] keys


i'd like to point out that even c++ has dropped support for trigraphs https://docs.microsoft.com/en-us/cpp/c-language/trigraphs?view=vs-2019

😉😛

OddBloke commented on cloudinit/config/cc_ssh_import_id.py:
```diff
@@ -56,13 +60,16 @@ def handle(_name, cfg, cloud, log, args):
+
+        # handle space separated list
+        if isinstance(import_ids, str):
+            import_ids = import_ids.split(" ")
```
Specifying `" "` here means that multiple spaces will result in empty string list entries:

```py
>>> "foo  bar".split(" ")
['foo', '', 'bar']
>>> "foo  bar".split()
['foo', 'bar']
```

and that non-ASCII spaces won't be correctly split on:

```py
>>> "foo\u2009bar".split(" ")
['foo\u2009bar']
>>> "foo\u2009bar".split()
['foo', 'bar']
```

blackboxsw replied:
> Specifying `" "` here means that multiple spaces will result in empty string list entries:
> 
> ```python
> >>> "foo  bar".split(" ")
> ['foo', '', 'bar']
> >>> "foo  bar".split()
> ['foo', 'bar']
> ```

Empty items won't matter because later that list is passed to util.uniq_merge which drops empties.
But agreed on non-ASCII whitespace.

And yes, this does complicate the json schema definition when written for this module, but it doesn't make it hard to validate, just a bit too "flexible" a schena instead of strict types. As, rharper metnioned, there are platform reasons, so I'm also ok with going this route.

raharper commented:
The use case is s390x console which can’t use [] and instead will use
kernel command line cc: ssh_import_id: lp:user1 lp:user2 end_cc

Kernel command line cannot use new lines.  Hence the patch

On Fri, Jan 24, 2020 at 21:28 Daniel Watkins <notifications@github.com>
wrote:

> *@OddBloke* requested changes on this pull request.
>
> I have one inline comment, but I'd also like to see some clearer
> justification of why we want to make this change. Generally speaking, we
> have been moving from using unstructured data to structured data for inputs
> (and making it easier to validate structured data with JSON schemas). This
> seems like a step in the opposite direction. I'm not *necessarily*
> opposed, but I do think we need to be clear about why we want to do this.
>
> I'll also note that you don't need to use [ or ] to list multiple keys:
>
> ssh_import_id:
>     - first_key
>     - second_key
>
> ------------------------------
>
> In cloudinit/config/cc_ssh_import_id.py
> <https://github.com/canonical/cloud-init/pull/187#discussion_r370799995>:
>
> >          else:
> -            try:
> -                import_ids = user_cfg['ssh_import_id']
> -            except Exception:
> -                log.debug("User %s is not configured for ssh_import_id", user)
> -                continue
> +            import_ids = user_cfg.get('ssh_import_id', [])
> +
> +        if not import_ids:
> +            continue
> +
> +        # handle space separated list
> +        if isinstance(import_ids, str):
> +            import_ids = import_ids.split(" ")
>
> Specifying " " here means that multiple spaces will result in empty
> string list entries:
>
> >>> "foo  bar".split(" ")
> ['foo', '', 'bar']>>> "foo  bar".split()
> ['foo', 'bar']
>
> and that non-ASCII spaces won't be correctly split on:
>
> >>> "foo\u2009bar".split(" ")
> ['foo\u2009bar']>>> "foo\u2009bar".split()
> ['foo', 'bar']
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/cloud-init/pull/187?email_source=notifications&email_token=ACTMSUZM2FRV4XTPFXW73WTQ7M6L7A5CNFSM4KLEA6T2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCTAKLOA#pullrequestreview-348169656>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTMSU4ELXDCCWANQHGLLRLQ7M6L7ANCNFSM4KLEA6TQ>
> .
>

OddBloke commented:
On Fri, Jan 24, 2020 at 11:31:50AM -0800, Ryan Harper wrote:
> The use case is s390x console which can’t use [] and instead will use
> kernel command line cc: ssh_import_id: lp:user1 lp:user2 end_cc
> 
> Kernel command line cannot use new lines.  Hence the patch

Why not base64-encode the input?

raharper commented:
User will type it in

On Fri, Jan 24, 2020 at 21:44 Daniel Watkins <notifications@github.com>
wrote:

> On Fri, Jan 24, 2020 at 11:31:50AM -0800, Ryan Harper wrote:
> > The use case is s390x console which can’t use [] and instead will use
> > kernel command line cc: ssh_import_id: lp:user1 lp:user2 end_cc
> >
> > Kernel command line cannot use new lines. Hence the patch
>
> Why not base64-encode the input?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/cloud-init/pull/187?email_source=notifications&email_token=ACTMSU3ELQA5TEQX5OJHO33Q7NAKXA5CNFSM4KLEA6T2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ34CJY#issuecomment-578273575>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTMSU3RURWKU4JSJ7E4S2DQ7NAKXANCNFSM4KLEA6TQ>
> .
>

OddBloke commented:
On Fri, Jan 24, 2020 at 12:07:10PM -0800, Ryan Harper wrote:
> User will type it in

<insert Picard facepalm here>

I still feel like I don't really understand _why_ any of this is
required.  Do we have a bug report or email or something that outlines
what the requirements here are?

raharper commented:
No,  just what I’m told from s390 install process this week.  I don’t think
this is terrible, we already convert a string to a single element list,
 usernames have to be with out a space so it doesn’t seem to be an issue
for parsing.



On Fri, Jan 24, 2020 at 22:25 Daniel Watkins <notifications@github.com>
wrote:

> On Fri, Jan 24, 2020 at 12:07:10PM -0800, Ryan Harper wrote:
> > User will type it in
>
> <insert Picard facepalm here>
>
> I still feel like I don't really understand _why_ any of this is
> required. Do we have a bug report or email or something that outlines
> what the requirements here are?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/cloud-init/pull/187?email_source=notifications&email_token=ACTMSU43HYAD3LCQCN672OLQ7NFB3A5CNFSM4KLEA6T2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ37PJI#issuecomment-578287525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTMSU6SKVMSWTWXTBS5B6LQ7NFB3ANCNFSM4KLEA6TQ>
> .
>

igalic commented:
> The use case is s390x console which can’t use [] and instead will use

i can't believe i was right.

raharper commented:
@OddBloke do you still need changes or more information?

xnox commented:
So on z/vm users need to upload a paramfile via ftp.
If the paramfile has [ it defaults to be treated as "binary" because that's not a valid encoding, in the default EBCDIC. Thus e.g. using static ipv6 addresses (because no dhcp / RA either) is really hard and error prone.
Note, quite often one cannot attach or setup local NoCloud or remote NoCloud datasource either, as the location of boot material is often on a separate network, from  access network. Thus I expect users will prove cc: datasource on the kernel commandline via paramfile, which is limited to 80char width and gets manged by the kernel a bit (i.e. new lines and whitespace sometimes get mangled).

So it would really really really nice for the z/VM admins to be able to specify something without [] on the kernel commandline to provide interactive access into the machines.

This is not a must, but will improve usability of server installer on z/VM by a 1000 times.

I only request to support "space separated" input on the kernel commandline string; all other datasources, i.e. noclound / nocloud-net / baked-in cloud.cfg should continue to only support structured data.

raharper commented:
> So it would really really really nice for the z/VM admins to be able to specify something without [] on the kernel commandline to provide interactive access into the machines.
> 
> This is not a must, but will improve usability of server installer on z/VM by a 1000 times.

Indeed.

> 
> I only request to support "space separated" input on the kernel commandline string; all other datasources, i.e. noclound / nocloud-net / baked-in cloud.cfg should continue to only support structured data.

@xnox how do you feel about URL encoding? 

```
>>> import urllib.parse
>>> urllib.parse.unquote('ssh_import_id: %5b user1, user2 %5d')
'ssh_import_id: [ user1, user2 ]'
```

raharper commented:
I've got the urlencoded version here, #275
